### PR TITLE
Improving workflow for dual frame optimization

### PR DESCRIPTION
**Pull request by timmintam**
The goal of this PR is to improve: 

1. the interface for the optimization of dual frames
2. performance of the optimization process
3. performance of the computation of the decomposition weights
4. memory usage of dual frames

### CODE REVIEW

File: povm_toolbox/library/randomized_projective_measurements.py [hunk 1]
```diff
@@ -170,7 +170,7 @@ def _build_qc(self) -> QuantumCircuit:
             qc.sx(qubit=i)
             qc.rz(np.pi - self._qc_theta[i], qubit=i)
             qc.sx(qubit=i)
-            qc.rz(3 * np.pi, qubit=i)
+            # qc.rz(3 * np.pi, qubit=i)
```
Comment (mrossinek) [inline povm_toolbox/library/randomized_projective_measurements.py:173]: Why is this necessary?

Comment (timmintam) [reply povm_toolbox/library/randomized_projective_measurements.py:173]: The POVM effects (and hence the outcome probabilities) are actually the same if we have this final gate or not, so I thought we could just remove it

Comment (mrossinek) [reply povm_toolbox/library/randomized_projective_measurements.py:178]: But in that case, let's outright remove the line of code :+1:

File: docs/how_tos/dual_optimizer.ipynb
Comment (mrossinek) [inline docs/how_tos/dual_optimizer.ipynb:1]: I think you forgot to finish the intro here.

Comment (mrossinek) [reply docs/how_tos/dual_optimizer.ipynb:1]: Also: the text around the measurement layout does not apply to this guide.

Comment (mrossinek) [reply docs/how_tos/dual_optimizer.ipynb:1]: And finally, maybe you can add some very brief explanations for each of the different post-processors.

File: docs/tutorials/getting_started.ipynb
Comment (mrossinek) [inline docs/tutorials/getting_started.ipynb:1]: Why did the output change here? Do we not fix the random seed in this notebook?

File: docs/tutorials/parametrized_circuit.ipynb
Comment (mrossinek) [inline docs/tutorials/parametrized_circuit.ipynb:1]: I think this notebook may also be missing a fixed random seed. Possibly more below, will not comment again.

File: povm_toolbox/post_processor/povm_post_processor.py [hunk 2]
```diff
@@ -56,59 +67,90 @@ def dual(self) -> BaseDUAL:
...
-        # second when we optimize it).
-        self._dual = self.dual.optimize(self.povm, **options)
+    def get_decomposition_weights(
+        self, observable: SparsePauliOp, outcome_idx: set[Any]
+    ) -> dict[Any, float]:
+        """Get the decomposition weights of ``observable`` into the elements of ``self.povm``.
+
+        Args:
+            observable: the observable to be decomposed into the POVM effects.
+            outcome_idx: set of labels indicating which decomposition weights are queried.
```
Comment (mrossinek) [inline povm_toolbox/post_processor/povm_post_processor.py:81]: I think this needs more explanation

File: povm_toolbox/post_processor/povm_post_processor.py [hunk 3]
```diff
@@ -56,59 +67,90 @@ def dual(self) -> BaseDUAL:
...
+
+        Returns:
+            An dictionary of decomposition weights.
```
Comment (mrossinek) [inline povm_toolbox/post_processor/povm_post_processor.py:84]:
```suggestion
            A dictionary of decomposition weights.
```

File: povm_toolbox/post_processor/povm_post_processor.py [hunk 4]
```diff
@@ -56,59 +67,90 @@ def dual(self) -> BaseDUAL:
...
         count = self.counts[loc]
+        shots = sum(count.values())
         # TODO: performance gains to be made when computing the omegas here ?
         # like storing the dict of computed omegas and updating the dict with the
         # missing values that were still never computed.
-        omegas = dict(self.dual.get_omegas(observable, set(count.keys())))  # type: ignore
+        omegas = self.get_decomposition_weights(observable, set(count.keys()))
+
+        exp_val = 0.0
+        pw2 = 0.0
```
Comment (mrossinek) [inline povm_toolbox/post_processor/povm_post_processor.py:144]: Please use a better variable name that is more self-describing

File: povm_toolbox/post_processor/ppp_empirical_frequencies.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+import numpy as np
+from qiskit.quantum_info import DensityMatrix, SparsePauliOp, Statevector
+
+from povm_toolbox.post_processor import POVMPostProcessor
+from povm_toolbox.quantum_info import ProductPOVM
+from povm_toolbox.quantum_info.product_dual import ProductDUAL
+
+
+class PPPEmpiricalFrequencies(POVMPostProcessor):
```
Comment (mrossinek) [inline povm_toolbox/post_processor/ppp_empirical_frequencies.py:23]: Does `PPP` stand for **P**OVM**P**ost**P**rocessor?

Not sure whether we need that in all these names... We don't do that with `ClassicalShadows` either (for example).
Maybe `EmpiricalFrequencies` is enough, given that this can only be imported from the `post_processor` module to begin with?

File: povm_toolbox/quantum_info/base_dual.py [hunk 6]
```diff
@@ -41,22 +42,15 @@ def get_omegas(
...
     @classmethod
     @abstractmethod
-    def build_dual_from_frame(cls, frame: BaseFrame) -> BaseDUAL:
+    def build_dual_from_frame(cls, frame: BaseFrame, alphas: tuple[Any] | None = None) -> BaseDUAL:
         """Construct a dual frame to another frame.
 
         Args:
             frame: The primal frame from which we will build the dual frame.
+            alphas: parameters of the local frame super-operators used to build
+                the local dual frames which form together the product dual frame.
```
Comment (mrossinek) [inline povm_toolbox/quantum_info/base_dual.py:53]: Please explain why this can be `None` and, if it is, what that implies (e.g. what default will be used).

File: povm_toolbox/quantum_info/product_dual.py [hunk 7]
```diff
@@ -73,38 +73,18 @@ def is_dual_to(self, frame: BaseFrame) -> bool:
...
         """Construct a dual frame to another frame.
 
         Args:
-            frame: The primal frame from which we will build the dual frame.
-            alphas: TODO.
+            frame: the primal frame from which we will build the dual frame.
+            alphas: parameters of the local frame super-operators used to build
+                the local dual frames which form together the product dual frame.
+                If None, the parameters are set as the traces of each local operator
+                in each of the primal frames.
```
Comment (mrossinek) [inline povm_toolbox/quantum_info/product_dual.py:87]: This should be aligned with the docstring o the base class (as much as possible).

File: test/library/test_randomized_projective_measurements.py [hunk 8]
```diff
@@ -78,16 +78,16 @@ def test_twirling(self):
         post_processor = POVMPostProcessor(pub_result)
 
         observable = SparsePauliOp(["ZI"], coeffs=[1.0])
-        exp_value, _ = post_processor.get_single_exp_value_and_std(observable)
+        exp_value, _ = post_processor.get_expectation_value(observable)
```
Comment (mrossinek) [inline test/library/test_randomized_projective_measurements.py:81]: Could you also assert the standard deviation, please?
In all of these cases.

**Review by mrossinek — COMMENTED** [review]
I still need to review the three new PPPs but submitting this review as is for now :+1:

File: docs/tutorials/getting_started.ipynb
Comment (timmintam) [reply docs/tutorials/getting_started.ipynb:1]: Yes it looks like some notebooks don't have their random seed(s) fixed. I will fix (😉) this issue.

File: test/sampler/test_povm_sampler_job.py [hunk 9]
```diff
@@ -76,7 +76,7 @@ def test_recover_job(self):
         pub_result = result[0]
         observable = SparsePauliOp(["II", "XX", "YY", "ZZ"], coeffs=[1, 1, -1, 1])
         post_processor = POVMPostProcessor(pub_result)
-        exp_value = post_processor.get_expectation_value(observable)
+        exp_value, _ = post_processor.get_expectation_value(observable)
```
Comment (mrossinek) [inline test/sampler/test_povm_sampler_job.py:79]: These should also assert the standard deviation (same below)

File: povm_toolbox/post_processor/state_optimal.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            NotImplementedError: if ``self.povm`` is not a :class:`povm_toolbox.quantum_info.product_povm.MultiQubitPOVM`
+                instance.
+        """
+        if not isinstance(self.povm, MultiQubitPOVM):
+            # TODO : implement for ProductPOVM
+            raise NotImplementedError(
+                "This method is only implemented for `povm_toolbox.quantum_info.product_povm.MultiQubitPOVM`."
+            )
+        alphas = tuple(self.povm.get_prob(state))  # type: ignore
+        self._dual = MultiQubitDUAL.build_dual_from_frame(self.povm, alphas=alphas)
```
Comment (mrossinek) [inline povm_toolbox/post_processor/state_optimal.py:49]: I think this is a prime candidate for a different implementation approach of the sort that we were discussing offline: rather than making subclasses of `POVMPostProcessor` which all just add a single method to set a custom `self._dual`, maybe we can expose a public setter for `self.dual` on the `POVMPostProcessor` and then provide a number of different functions which the user calls to set this dual.

Here, this would be a particularly simple example:
```python
def optimal_dual_from_state(state, povm):
    alphas = tuple(povm.get_prob(state))
    return MultiQubitDUAL.build_dual_from_frame(povm, alphas=alphas)

post_processor = POVMPostProcessor(result)
post_processor.dual = optimal_dual_from_state(state, povm)
```

Comment (mrossinek) [reply povm_toolbox/post_processor/state_optimal.py:49]: The same will apply to the other subclasses, but I will review them in terms of their content, not w.r.t. this design aspect.

File: povm_toolbox/post_processor/state_marginal.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            raise NotImplementedError(
+                "This method is only implemented for `povm_toolbox.quantum_info.product_povm.ProductPOVM`."
+            )
+        axes = np.arange(len(self.povm.sub_systems), dtype=int)
+        joint_prob: np.ndarray = self.povm.get_prob(state)  # type: ignore
+        alphas = []
+        for qubit_idx in self.povm.sub_systems:
+            marg_prob = joint_prob.sum(axis=tuple(np.delete(axes, [qubit_idx])))
+            if np.any(np.absolute(marg_prob) < 1e-5):
+                marg_prob += 1e-5
```
Comment (mrossinek) [inline povm_toolbox/post_processor/state_marginal.py:53]: Should this hard-coded `1e-5` be exposed as a threshold on this function?

File: povm_toolbox/post_processor/empirical_frequencies.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        for outcome, count in counts.items():
+            for i, k_i in enumerate(outcome):
+                marginals[i][k_i] += count / shots
+
+        alphas = []
+        # Computing alphas for each subsystem
+        for i, sub_system in enumerate(self.povm.sub_systems):
+            sub_povm = self.povm[sub_system]
+            dim = sub_povm.dimension
+            ansatz_state = DensityMatrix(np.eye(dim) / dim) if ansatz is None else ansatz[i]
```
Comment (mrossinek) [inline povm_toolbox/post_processor/empirical_frequencies.py:93]: The length of `ansatz` should be asserted to match the length of `povm_sub_systems`

File: povm_toolbox/post_processor/empirical_frequencies.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            sub_bias = (
+                sub_povm.n_outcomes
+                if bias is None
+                else (bias[i] if isinstance(bias, list) else bias)
```
Comment (mrossinek) [inline povm_toolbox/post_processor/empirical_frequencies.py:97]: Also, this length should be asserted.

Comment (mrossinek) [reply povm_toolbox/post_processor/empirical_frequencies.py:97]: Given that one can apply the same bias to all sub-systems, why don't we allow that for the ansatz?

**Review by mrossinek — COMMENTED** [review]
Some more comments after my earlier review

File: povm_toolbox/post_processor/dual_from_empirical_frequencies.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        NotImplementedError: if ``povm_post_processor.povm`` is not a
+            :class:`povm_toolbox.quantum_info.product_povm.ProductPOVM`
+            instance.
+        ValueError: if `loc` is None and that the POVM post-processor stores more
+            than one counter (i.e., multiple sets of parameter values were
+            supplied to the sampler in a single pub).
+        ValueError: if `bias` is a list but its length does not match the number
+            of local POVMs forming the product POVM ``povm_post_processor.povm``.
+        ValueError: if `ansatz` is a list but its length does not match the number
+            of local POVMs forming the product POVM ``povm_post_processor.povm``.
```
Comment (mrossinek) [inline povm_toolbox/post_processor/dual_from_empirical_frequencies.py:71]:
```suggestion
        ValueError: if ``loc`` is None and that the POVM post-processor stores more
            than one counter (i.e., multiple sets of parameter values were
            supplied to the sampler in a single pub).
        ValueError: if ``bias`` is a list but its length does not match the number
            of local POVMs forming the product POVM ``povm_post_processor.povm``.
        ValueError: if ``ansatz`` is a list but its length does not match the number
            of local POVMs forming the product POVM ``povm_post_processor.povm``.
```

**Review by mrossinek — COMMENTED** [review]
Some more comments and modulo the unittests this is good to go :+1:

File: povm_toolbox/post_processor/dual_from_empirical_frequencies.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    """Return the dual frame of `povm` based on the frequencies of the sampled outcomes.
+
+    Given outcomes sampled from a product POVM, each local dual frame is parametrized
+    with the alpha-parameters set as the marginal outcome frequencies. For stability,
+    the (local) empirical frequencies can be biased towards the (marginal) outcome
+    probabilities of an ansatz state.
+
+    Args:
+        povm_post_processor: the :class:`.POVMPostProcessor` object from which to
+            extract the POVM and the empirical frequencies to build the dual frame.
```
[the remark this answers is not in the document]
Comment (timmintam) [reply povm_toolbox/post_processor/dual_from_empirical_frequencies.py:42]: Yeah I wasn't so sure either... 

- The advantage of `povm` and `counts` is that we can get rid of the cumbersome `loc` argument (the user has to do the work himself for extracting the correct counts). Also is is more aligned with the other dual constructors as you first have to feed in a POVM.
- The disadvantage is that there's no guarantee that the `counts` comes from the correct `povm`. Also both arguments will likely be anyway extracted from the post-processor.

File: povm_toolbox/post_processor/dual_from_state.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+from typing import cast
+
+import numpy as np
+from qiskit.quantum_info import DensityMatrix, SparsePauliOp, Statevector
+
+from povm_toolbox.quantum_info import BaseDUAL, BasePOVM, MultiQubitPOVM
+from povm_toolbox.quantum_info.multi_qubit_dual import MultiQubitDUAL
+
+
+def optimal_dual_from_state(
```
Comment (mrossinek) [inline povm_toolbox/post_processor/dual_from_state.py:24]: Why did you rename this to `optimize_dual_from_state` rather than just `dual_from_state`?
There is no "optimization" performed here as this is a one-shot computation of the dual from the provided state.

File: test/post_processor/test_dual_from_state.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+from unittest import TestCase
+
+import numpy as np
+from povm_toolbox.library import ClassicalShadows, LocallyBiasedClassicalShadows
+from povm_toolbox.post_processor import optimal_dual_from_state
+from qiskit.quantum_info import DensityMatrix, random_density_matrix
+
+
+class TestOptimalDualFromState(TestCase):
```
Comment (mrossinek) [inline test/post_processor/test_dual_from_state.py:21]:
```suggestion
class TestDualFromState(TestCase):
```
Assuming you agree with my comment above.

File: test/post_processor/test_dual_from_state.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    def test_not_implemented(self):
+        """Test that errors are correctly raised."""
+        prod_povm = ClassicalShadows(n_qubit=2).definition()
+        state = DensityMatrix(np.eye(4) / 4)
+        with self.assertRaises(NotImplementedError):
+            _ = optimal_dual_from_state(prod_povm, state)
+
+    def test_optimal_dual(self):
+        """Test that the method constructs a valid dual."""
+        povm = LocallyBiasedClassicalShadows(
```
Comment (mrossinek) [inline test/post_processor/test_dual_from_state.py:33]: It is a bit confusing that you once use CS and once LB-CS even though the actual difference that you are testing is `ProductPOVM` vs `MultiQubitPOVM`.
Can you amend this slightly to not give this "wrong" message.

File: test/post_processor/test_dual_from_state.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Tests for the MultiQubitPOVM class."""
```
Comment (mrossinek) [inline test/post_processor/test_dual_from_state.py:11]: This is outdated (same in the other files).

File: test/post_processor/test_dual_from_marginal_probabilities.py [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    Statevector,
+)
+
+
+class TestDualFromMarginalProbabilities(TestCase):
+    """Test that we can construct optimal dual of a POVM from marginal probabilities."""
+
+    def test_not_implemented(self):
+        """Test that errors are correctly raised."""
+        povm = ClassicalShadows(n_qubit=1).definition()[(0,)]
```
Comment (mrossinek) [inline test/post_processor/test_dual_from_marginal_probabilities.py:40]: The same comment as before but here mixing CS and RPMs.
Maybe you can update the not-implemented test case in all places to directly build a non-working `BasePOVM` (product or multi depending on the case) to show that is the culprit and not a particular library POVM.

File: test/post_processor/test_dual_from_state.py [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+            _ = optimal_dual_from_state(prod_povm, state)
+
+    def test_optimal_dual(self):
+        """Test that the method constructs a valid dual."""
+        povm = LocallyBiasedClassicalShadows(
+            n_qubit=1, bias=np.array([0.3, 0.6, 0.1])
+        ).definition()[(0,)]
+        state = random_density_matrix(2, seed=12)
+        dual = optimal_dual_from_state(povm, state)
+        self.assertTrue(dual.is_dual_to(povm))
```
Comment (mrossinek) [inline test/post_processor/test_dual_from_state.py:38]: Can you also assert some expectation value and standard deviation here, just like you do in the other test cases?
It may also be nice to assert it before and after setting the optimal dual to showcase the difference.

File: test/post_processor/test_dual_from_empirical_frequencies.py [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+        self.assertAlmostEqual(std, 0.40472771240833644)
+
+        post_processor.dual = dual_from_empirical_frequencies(
+            povm_post_processor=post_processor,
+            bias=[6, 6],
+            ansatz=SparsePauliOp(["I"], coeffs=np.array([0.5])),
+        )
+        exp_value, std = post_processor.get_expectation_value(observable)
+        self.assertAlmostEqual(exp_value, 0.9128662937761666)
+        self.assertAlmostEqual(std, 0.40472771240833644)
```
Comment (mrossinek) [inline test/post_processor/test_dual_from_empirical_frequencies.py:94]: Please use sub-tests when you want to perform multiple assertion blocks: https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

File: docs/how_tos/dual_optimizer.ipynb
Comment (mrossinek) [inline docs/how_tos/dual_optimizer.ipynb:1]: Nice, this is very good now! :+1: 

One last correction:
> More precisely, we replace the the marginal outcome probabilities

Please remove one of the two `the` articles here.

**Review by mrossinek — COMMENTED** [review]
Almost done! Some minor feedback on the unittests and one naming discussion

File: povm_toolbox/post_processor/dual_from_state.py (hunk 16 above)
Comment (timmintam) [reply povm_toolbox/post_processor/dual_from_state.py:24]: It's `optimal` (not optimized) because we have the analytical formula for the optimal dual frame if we know the state. But I agree, maybe we should just align with the other names. Let me know what you think!

Comment (mrossinek) [reply povm_toolbox/post_processor/dual_from_state.py:24]: Oh I totally misread that.. :sweat_smile: 

But yeh, let's keep it as `dual_from_state`. You can document that this is the optimal one (plus, users will likely notice when they realize that they need to input the state into this..)

File: test/post_processor/test_dual_from_empirical_frequencies.py [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+class TestDualFromEmpiricalFrequencies(TestCase):
+    """Test that we can construct optimal dual of a POVM from empirical frequencies."""
+
+    def test_not_implemented(self):
+        """Test that errors are correctly raised."""
+        _ = ClassicalShadows(n_qubit=1).definition()[(0,)]
+        _ = DensityMatrix(np.eye(2) / 2)
+        with self.assertRaises(NotImplementedError):
+            # TODO
+            raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (timmintam) [reply test/post_processor/test_dual_from_empirical_frequencies.py]: As discussed online, I will remove this test for now.

**Review by mrossinek — APPROVED** [review]
Thanks a lot! :+1:
